File: workstation/src/test/java/io/xj/model/util/CsvUtilsTest.java
// Copyright (c) XJ Music Inc. (https://xj.io) All Rights Reserved.
package io.xj.model.util;

import java.util.Map;

import org.junit.jupiter.api.Test;

import java.util.Arrays;
import java.util.List;
import java.util.UUID;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

public class CsvUtilsTest {
  @Test
  public void split() {
    assertEquals(
      Arrays.asList("one", "two", "three"),
      CsvUtils.split("one,two,three"));
  }

  @Test
  public void splitUUIDs() {
    assertEquals(
      Arrays.asList(
        UUID.fromString("e33e4c97-49f5-4fa8-85bf-3f0618eca351"),
        UUID.fromString("e0e6c13c-9f84-4641-b4e5-796af7100864"),
        UUID.fromString("cc7a6a7a-113e-44bb-a892-29a7a5610945")
      ),
      CsvUtils.splitUUIDs("e33e4c97-49f5-4fa8-85bf-3f0618eca351,e0e6c13c-9f84-4641-b4e5-796af7100864,cc7a6a7a-113e-44bb-a892-29a7a5610945"));
  }

  @Test
  public void split_null() {
    assertEquals(
      List.of(),
      CsvUtils.split(null));
  }

  @Test
  public void split_empty() {
    assertEquals(
      List.of(),
      CsvUtils.split(""));
  }

  @Test
  public void split_alsoTrims() {
    assertEquals(
      Arrays.asList("one", "two", "three"),
      CsvUtils.split("one, two, three"));
  }

  @Test
  public void splitProperSlug() {
    assertEquals(
      Arrays.asList("One", "Two", "Three"),
      CsvUtils.splitProperSlug("one,two,three"));
  }

  @Test
  public void join() {
    assertEquals(
      "one, two, three",
      CsvUtils.join(Arrays.asList("one", "two", "three")));
  }

  @Test
  public void from_keyValuePairs() {
    var result = CsvUtils.from(Map.of("one", "1", "two", "2", "three", "3"));

    // can't use this because map has no guaranteed order: assertEquals("one=1, two=2, three=3");
    assertEquals(21, result.length());
    assertTrue(result.contains("one=1"));
    assertTrue(result.contains("two=2"));
    assertTrue(result.contains("three=3"));
  }

  @Test
  public void prettyFrom() {
    assertEquals(
      "One, Two, or Three",
      CsvUtils.prettyFrom(Arrays.asList("One", "Two", "Three"), "or"));
  }
}

File: engine/test/util/CsvUtilsTest.cpp
// Copyright (c) XJ Music Inc. (https://xjmusic.com) All Rights Reserved.

#include <gtest/gtest.h>

#include <vector>
#include <string>

#include "xjmusic/util/CsvUtils.h"

using namespace XJ;


TEST(CsvUtilsTest, split) {
  const std::vector<std::string> expected = {"one", "two", "three"};

  const auto actual = CsvUtils::split("one,two,three");

  ASSERT_EQ(expected, actual);
}


TEST(CsvUtilsTest, split_empty) {
  ASSERT_EQ(std::vector<std::string>{}, CsvUtils::split(""));
}


TEST(CsvUtilsTest, split_alsoTrims) {
  const std::vector<std::string> expected = {"one", "two", "three"};

  const auto actual = CsvUtils::split("one, two, three");

  ASSERT_EQ(expected, actual);
}


TEST(CsvUtilsTest, splitProperSlug) {
  const std::vector<std::string> expected = {"One", "Two", "Three"};

  const auto actual = CsvUtils::splitProperSlug("one,two,three");

  ASSERT_EQ(expected, actual);
}


TEST(CsvUtilsTest, join) {
  const std::string expected = "one, two, three";

  const auto actual = CsvUtils::join({"one", "two", "three"});

  ASSERT_EQ(expected, actual);
}


TEST(CsvUtilsTest, from_keyValuePairs) {
  const auto result = CsvUtils::from({
                                   {"one",   "1"},
                                   {"two",   "2"},
                                   {"three", "3"}
                               });

  ASSERT_EQ(21, result.length());
  ASSERT_TRUE(result.find("one=1") != std::string::npos);
  ASSERT_TRUE(result.find("two=2") != std::string::npos);
  ASSERT_TRUE(result.find("three=3") != std::string::npos);
}


TEST(CsvUtilsTest, prettyFrom) {
  const std::string expected = "One, Two, or Three";

  const auto actual = CsvUtils::prettyFrom({"One", "Two", "Three"}, "or");

  ASSERT_EQ(expected, actual);
}

TEST(StringUtilsTest, toProperCsvAnd) {
  EXPECT_EQ("One", CsvUtils::toProperCsvAnd({"One"}));
  EXPECT_EQ("One and Two", CsvUtils::toProperCsvAnd({"One", "Two"}));
  EXPECT_EQ("One, Two, and Three", CsvUtils::toProperCsvAnd({"One", "Two", "Three"}));
}

TEST(StringUtilsTest, toProperCsvOr) {
  EXPECT_EQ("One", CsvUtils::toProperCsvOr({"One"}));
  EXPECT_EQ("One or Two", CsvUtils::toProperCsvOr({"One", "Two"}));
  EXPECT_EQ("One, Two, or Three", CsvUtils::toProperCsvOr({"One", "Two", "Three"}));
}

TEST(StringUtilsTest, toProperCsv) {
  EXPECT_EQ("One", CsvUtils::toProperCsv({"One"}, "or maybe"));
  EXPECT_EQ("One or maybe Two", CsvUtils::toProperCsv({"One", "Two"}, "or maybe"));
  EXPECT_EQ("One, Two, or maybe Three", CsvUtils::toProperCsv({"One", "Two", "Three"}, "or maybe"));
}